webhook/main.py

```diff
@@ -12,33 +12,333 @@
+    index_pages(index_id, filename, pages)
+
+    print(f"🔍 {event_id}: Generating Q&As with model ({len(pages)} pages)")
+    with multiprocessing.Pool(len(pages)) as pool:
+        event_pages = [
+            {"filename": filename, "page_number": i, "text": page}
+            for i, page in enumerate(pages)
+        ]
+        results = pool.map(process_page, event_pages)
+        entries = list(
+            itertools.chain.from_iterable(results)
+        )
+    
+    print(f"🗃️ {event_id}: Saving Q&As to Firestore ({len(entries)} entries)")
+    for entry in entries:
+        doc = db.document("dataset", entry["question"].replace("/", " "))
+        if doc.get().exists:
+            doc.update(entry)
+        else:
+            doc.create(entry)
+
+    print(f"📝 {event_id}: Writing tuning dataset: gs://{output_bucket}/dataset.jsonl")
+    dataset_size = write_tuning_dataset(db, output_bucket)
+    print(f"✅ {event_id}: Done! {dataset_size=}")
+
+
+def process_page(event_page: dict) -> list[dict[str, str]]:
+    """Generate questions and answers for a single page of a document.
+
+    Args:
+        event_page: Dictionary containing the filename, page number, and text of the pages.
+
+    Returns: Dictionaries containing the questions and answers.
+    """
+    filename = event_page["filename"]
+    page_number = event_page["page_number"]
+    text = event_page["text"]
+    entries = generate_questions(text)
+    try:
+        return [
+            {
+                "question": entry["question"],
+                "answer": entry["answer"],
+                "filename": filename,
+                "page_number": page_number,
+            }
+            for entry in entries
+        ]
+    except KeyError:
+        logging.exception(f"Q&A generation failed: {entries}", stack_info=True)
+        return []
+
+
+def get_document_text(
+    gcs_uri: str,
+    mime_type: str,
+    processor_id: str,
+) -> list[str]:
+    """Perform Optical Character Recognition (OCR) with Document AI on a Cloud Storage files.
+
+    For more information, see:
+        https://cloud.google.com/document-ai/docs/process-documents-ocr
+
+    Args:
+        gcs_uri: GCS URI of the document file.
+        mime_type: MIME type of the document file.
+        processor_id: ID of the Document AI processor.
+
+    Returns: A list of the text in each page of the document.
+    """
+    # You must set the `api_endpoint` if you use a location other than "us".
+    client = documentai.DocumentProcessorServiceClient(
+        client_options=ClientOptions(
+            api_endpoint=f"{DOCAI_LOCATION}-documentai.googleapis.com"
+        )
+    )
+    response = client.process_document(
+        request=documentai.ProcessRequest(
+            name=processor_id,
+            gcs_document=documentai.GcsDocument(
+                gcs_uri=gcs_uri,
+                mime_type=mime_type,
+            ),
+        ),
+    )
+    return [
```

Comment:
We can replace this whole `return` with this:

```py
page_segments = [
    [
        (segment.start_index, segment.end_index)
        for segment in page.layout.text_anchor.text_segments
    ]
    for page in response.document.pages
]
return [
    "\n".join(
        response.document.text[start_index:end_index]
        for start_index, end_index in segments
    )
    for segments in page_segments
]
```

Reply:
Alternatively, this could also be an option, whatever you find easier to read.

```py
def get_pages(document: documentai.Document) -> Iterator[str]:
    for page in document.pages:
        segments = [(segment.start_index, segment.end_index) for segment in page.layout.text_anchor.text_segments]
        yield "\n".join(document.text[start_index:end_index] for start_index, end_index in segments)

return list(get_pages(response.document))
```